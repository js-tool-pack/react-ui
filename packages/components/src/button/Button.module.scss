@import '../scss.variable';

:global {
  .#{$PREFIX}button {
    --t-btn-main-color: black;
    --t-btn-color: white;
    --t-btn-bg-color: unset;
    --t-btn-border-style: solid;
    --t-btn-border-color: unset;
    --t-btn-border-width: 0;
    --t-btn-radius: var(--t-redis);
    --t-btn-padding: 5px;

    --t-btn-hover-color: white;
    --t-btn-hover-border-color: var(--t-btn-hover-color);
    --t-btn-hover-bg-color: unset;

    --t-btn-danger-color: red;
    --t-btn-danger-hover-color: #ff7875;

    position: relative;
    border-radius: var(--t-btn-radius);
    padding: 0 var(--t-btn-padding);
    height: var(--t-size);
    outline: 0;
    cursor: pointer;
    border: var(--t-btn-border-width) var(--t-btn-border-style)
      var(--t-btn-border-color);
    background-color: var(--t-btn-bg-color);
    color: var(--t-btn-color);
    transition: all var(--t-transition-duration) ease-in-out;

    &:not([disabled]):hover {
      --t-btn-color: var(--t-btn-hover-color);
      --t-btn-bg-color: var(--t-btn-hover-bg-color);
      --t-btn-border-color: var(--t-btn-hover-border-color);
    }

    &[disabled] {
      cursor: not-allowed;
    }
    &.type-primary {
      --t-btn-main-color: var(--t-primary-bg-color);
      --t-btn-bg-color: var(--t-primary-bg-color);
      --t-btn-hover-bg-color: var(--t-primary-hover-bg-color);
      &[disabled] {
        --t-btn-bg-color: var(--t-primary-disabled-bg-color);
      }
    }
    &.type-success {
      --t-btn-main-color: #67c23a;
      --t-btn-bg-color: #67c23a;
      --t-btn-hover-bg-color: #a9ea89;
      &[disabled] {
        --t-btn-bg-color: #c5f6ac;
      }
    }
    &.type-warning {
      --t-btn-main-color: #e6a23c;
      --t-btn-bg-color: #e6a23c;
      --t-btn-hover-bg-color: #eebe77;
      &[disabled] {
        --t-btn-bg-color: #f3cf9a;
      }
    }
    &.type-info {
      --t-btn-main-color: #909399;
      --t-btn-bg-color: #909399;
      --t-btn-hover-bg-color: #b1b3b8;
      &[disabled] {
        --t-btn-bg-color: #d5d7da;
      }
    }
    &.type-danger {
      --t-btn-main-color: #f56c6c;
      --t-btn-bg-color: #f56c6c;
      --t-btn-hover-bg-color: #f89898;
      &[disabled] {
        --t-btn-bg-color: #f89898;
      }
    }

    &.plain {
      --t-btn-border-width: 1px;
      --t-btn-border-style: solid;
      --t-btn-border-color: var(--t-btn-bg-color);
      --t-btn-color: var(--t-btn-bg-color);

      --t-btn-hover-border-color: var(--t-btn-hover-bg-color);
      --t-btn-hover-color: var(--t-btn-hover-bg-color);
      background-color: unset;
    }

    &.plain-dashed {
      --t-btn-border-width: 1px;
      --t-btn-border-style: dashed;
      --t-btn-border-color: var(--t-btn-bg-color);
      --t-btn-color: var(--t-btn-bg-color);

      --t-btn-hover-border-color: var(--t-btn-hover-bg-color);
      --t-btn-hover-color: var(--t-btn-hover-bg-color);
      background-color: unset;
    }
    &.plain-text {
      --t-btn-border-width: 0;
      --t-btn-border-style: dashed;
      --t-btn-border-color: var(--t-btn-bg-color);
      --t-btn-color: var(--t-btn-bg-color);

      --t-btn-hover-border-color: var(--t-btn-hover-bg-color);
      --t-btn-hover-color: var(--t-btn-hover-bg-color);
      background-color: unset;

      &:not([disabled]):active {
        transition: all calc(var(--t-transition-duration) / 3) ease-out;
        transform: translateY(-5px);
      }
    }

    &.type-default {
      --t-btn-main-color: var(--t-primary-bg-color);
      --t-btn-border-width: 1px;
      --t-btn-color: var(--t-text-color);
      --t-btn-border-color: black;
      --t-btn-hover-border-color: var(--t-primary-bg-color);
      --t-btn-hover-color: var(--t-primary-bg-color);
      &[disabled] {
        --t-btn-color: #cecece;
        --t-btn-border-color: #cecece;
      }
      &.plain-text {
        --t-btn-border-width: 0;
      }
    }
    &.#{$PREFIX}size-lg {
      --t-btn-padding: 10px;
    }
    &.shape-round,
    &.shape-circle {
      --t-btn-radius: calc(var(--t-size) / 2);
    }
    &.shape-circle {
      width: var(--t-size);
      //overflow: hidden;
      //text-overflow: ellipsis;
      //white-space: nowrap;
      &.#{$PREFIX}size-lg {
        --t-btn-padding: 5px;
      }
    }

    .btn-wave {
      position: absolute;
      left: 0;
      right: 0;
      top: 0;
      bottom: 0;
      border-radius: inherit;
      animation-duration: 0.5s;
      animation-iteration-count: 1;
    }
    .btn-wave-active {
      animation-name: btn-wave;
    }
  }

  @keyframes :global(btn-wave) {
    0% {
      box-shadow: 0 0 0.5px 0 var(--t-btn-main-color);
      opacity: 0.65;
    }

    100% {
      box-shadow: 0 0 0.5px 4.5px var(--t-btn-main-color);
      opacity: 0;
    }
  }
}
