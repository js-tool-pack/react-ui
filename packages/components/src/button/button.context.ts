import { createContext } from 'react';
import type { ButtonProps } from './button.types';

export const ButtonContext = createContext<ButtonProps>({});
