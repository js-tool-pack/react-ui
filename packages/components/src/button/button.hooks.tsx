import React, { useEffect, useMemo, useRef, useState } from 'react';
import { getClassNames } from '@tool-pack/basic';
import { CollapseTransition } from '../collapse-transition';
import { Icon } from '../icon';
import { Loading as LoadingIcon } from '@pkg/icons';
import type { ButtonProps } from './button.types';

export function useBtnWave(
  rootClass: string,
): readonly [boolean | React.ReactElement, () => void] {
  const [active, setActive] = useState(false);
  const waveRef = useRef<HTMLElement>(null);
  const activationTime = useRef(0);

  const activateWave = () => {
    if (active) {
      if (Date.now() - activationTime.current > 1500) setActive(false);
      return;
    }
    activationTime.current = Date.now();
    setActive(true);
  };

  useEffect(() => {
    const el = waveRef.current;
    if (!active || !el) return;
    const cancel = () => {
      el.removeEventListener('animationend', handler);
      el.removeEventListener('animationcancel', handler);
    };
    const handler = () => {
      setActive(false);
      cancel();
    };

    if (active && el) {
      el.addEventListener('animationend', handler);
      el.addEventListener('animationcancel', handler);
    }

    return handler;
  }, [active]);

  const wave = active && (
    <span
      ref={waveRef}
      className={getClassNames(`${rootClass}__wave`, {
        [`${rootClass}__wave--active`]: active,
      })}></span>
  );

  return [wave, activateWave];
}

export function useBtnIcon(
  rootClass: string,
  icon: ButtonProps['icon'],
  loading: ButtonProps['loading'],
): React.ReactElement {
  return useMemo(() => {
    const className = `${rootClass}__icon`;
    const Loading = (
      <Icon className={className} key="loading-icon">
        <LoadingIcon />
      </Icon>
    );

    if (icon)
      return loading ? Loading : <Icon className={className}>{icon}</Icon>;
    return <CollapseTransition width>{loading && Loading}</CollapseTransition>;
  }, [loading, icon]);
}
