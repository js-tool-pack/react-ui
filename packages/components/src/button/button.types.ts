import type { PropsBase, Size } from '@pkg/shared';
import React from 'react';

type BTN = React.ButtonHTMLAttributes<HTMLButtonElement>;
export interface ButtonProps
  extends PropsBase<HTMLButtonElement>,
    // 因为 button 的 className、onClick、style 比较常用，所以放置在外面
    Pick<BTN, 'className' | 'disabled' | 'onClick' | 'style'> {
  type?: 'default' | 'primary' | 'success' | 'warning' | 'danger' | 'info';
  attrs?: Partial<React.ButtonHTMLAttributes<HTMLButtonElement>>;
  shape?: 'default' | 'circle' | 'round' | 'none';
  plain?: 'dashed' | boolean | 'text';
  icon?: React.ReactElement | boolean;
  htmlType?: BTN['type'];
  rightIcon?: boolean;
  loading?: boolean;
  size?: Size;
}
export interface ButtonGroupProps extends PropsBase<HTMLElement> {
  size?: ButtonProps['size'];
}
