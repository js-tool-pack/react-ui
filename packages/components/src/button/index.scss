@use '../namespace' as Name;
@use '../scss.variable' as Var;

$ct: Name.$collapse-transition;
$r: Name.$button;
:root {
  --t-btn-wave-opcity: 0.6;
  &.dark {
    --t-btn-wave-opcity: 1;
  }
}

.#{$r} {
  --t-btn-main-color: var(--t-text-color);
  --t-btn-color: white;
  --t-btn-bg-color: unset;
  --t-btn-border-style: solid;
  --t-btn-border-color: unset;
  --t-btn-border-width: 0;
  --t-btn-radius: var(--t-radius);
  --t-btn-padding: 12px;
  --t-btn-icon-space: 6px;

  // ------
  --t-btn-hover-color: white;
  --t-btn-hover-border-color: var(--t-btn-hover-color);
  --t-btn-hover-bg-color: unset;

  // ------
  --t-btn-danger-color: red;
  --t-btn-danger-hover-color: #ff7875;

  position: relative;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  padding: 0 var(--t-btn-padding);
  height: var(--t-size);
  font-size: 1em;
  font-weight: 400;
  border: var(--t-btn-border-width) var(--t-btn-border-style)
    var(--t-btn-border-color);
  border-radius: var(--t-btn-radius);
  color: var(--t-btn-color);
  background-color: var(--t-btn-bg-color);
  outline: 0;
  user-select: none;
  cursor: pointer;
  transition-property: background-color, width, height, color, padding-top,
    padding-left, padding-right, padding-bottom, opacity;

  // transition-property: all;
  transition-timing-function: ease-in-out;
  transition-duration: var(--t-transition-duration);
  & > span {
    display: inline-flex;
    align-items: center;
  }
  &:not([disabled]):hover {
    --t-btn-color: var(--t-btn-hover-color);
    --t-btn-bg-color: var(--t-btn-hover-bg-color);
    --t-btn-border-color: var(--t-btn-hover-border-color);
  }
  &[disabled] {
    cursor: not-allowed;
  }
  &.#{$r}--type-primary {
    --t-btn-main-color: var(--t-primary-bg-color);
    --t-btn-bg-color: var(--t-primary-bg-color);
    --t-btn-hover-bg-color: var(--t-primary-hover-bg-color);
    &[disabled] {
      --t-btn-bg-color: var(--t-primary-disabled-bg-color);
    }
  }
  &.#{$r}--type-success {
    --t-btn-main-color: #67c23a;
    --t-btn-bg-color: #67c23a;
    --t-btn-hover-bg-color: #a9ea89;
    &[disabled] {
      --t-btn-bg-color: #c5f6ac;
    }
  }
  &.#{$r}--type-warning {
    --t-btn-main-color: #e6a23c;
    --t-btn-bg-color: #e6a23c;
    --t-btn-hover-bg-color: #eebe77;
    &[disabled] {
      --t-btn-bg-color: #f3cf9a;
    }
  }
  &.#{$r}--type-info {
    --t-btn-main-color: #909399;
    --t-btn-bg-color: #909399;
    --t-btn-hover-bg-color: #b1b3b8;
    &[disabled] {
      --t-btn-bg-color: #d5d7da;
    }
  }
  &.#{$r}--type-danger {
    --t-btn-main-color: #f56c6c;
    --t-btn-bg-color: #f56c6c;
    --t-btn-hover-bg-color: #f89898;
    &[disabled] {
      --t-btn-bg-color: #f89898;
    }
  }
  &.#{$r}--plain {
    --t-btn-border-width: 1px;
    --t-btn-border-style: solid;
    --t-btn-border-color: var(--t-btn-bg-color);
    --t-btn-color: var(--t-btn-bg-color);
    --t-btn-hover-border-color: var(--t-btn-hover-bg-color);
    --t-btn-hover-color: var(--t-btn-hover-bg-color);

    background-color: unset;
  }
  &.#{$r}--plain-dashed {
    --t-btn-border-width: 1px;
    --t-btn-border-style: dashed;
    --t-btn-border-color: var(--t-btn-bg-color);
    --t-btn-color: var(--t-btn-bg-color);
    --t-btn-hover-border-color: var(--t-btn-hover-bg-color);
    --t-btn-hover-color: var(--t-btn-hover-bg-color);

    background-color: unset;
  }
  &.#{$r}--plain-text {
    --t-btn-border-width: 0;
    --t-btn-border-style: dashed;
    --t-btn-border-color: var(--t-btn-bg-color);
    --t-btn-color: var(--t-btn-bg-color);
    --t-btn-hover-border-color: var(--t-btn-hover-bg-color);
    --t-btn-hover-color: var(--t-btn-hover-bg-color);

    background-color: unset;
    &:not([disabled]):active > * {
      transition: transform calc(var(--t-transition-duration) / 3) ease-out;
      transform: translateY(-5px);
    }
  }
  &.#{$r}--type-default {
    --t-btn-main-color: var(--t-primary-bg-color);
    --t-btn-border-width: 1px;
    --t-btn-color: var(--t-text-color);
    --t-btn-border-color: var(--t-btn-color);
    --t-btn-hover-border-color: var(--t-primary-bg-color);
    --t-btn-hover-color: var(--t-primary-bg-color);
    &[disabled] {
      --t-btn-color: #cecece;
      --t-btn-border-color: #cecece;
    }
    &.#{$r}--plain-text {
      --t-btn-border-width: 0;
    }
  }
  &.#{Var.$class-size-lg} {
    --t-btn-padding: 16px;
    --t-btn-icon-space: 8px;

    font-size: 16px;
  }
  &.#{Var.$class-size-sm} {
    --t-btn-padding: 6px;
    --t-btn-icon-space: 3px;
  }
  &.#{$r}--shape-round,
  &.#{$r}--shape-circle {
    --t-btn-radius: calc(var(--t-size) / 2);
  }
  &.#{$r}--shape-circle {
    width: var(--t-size);

    // overflow: hidden;
    // text-overflow: ellipsis;
    // white-space: nowrap;
    &.#{Var.$class-size-lg} {
      --t-btn-padding: 5px;
    }
  }
  &.#{$r}--shape-none {
    --t-btn-radius: 0;
  }
  .#{$r}__wave {
    position: absolute;
    inset: 0;
    border-radius: inherit;
    animation-duration: 0.5s;
    animation-iteration-count: 1;
  }

  .#{$r}__wave--active {
    animation-name: btn-wave;
  }
  &--icon-l {
    &:not(.#{$r}--icon-only) > .#{$r}__icon {
      margin-inline-end: var(--t-btn-icon-space);
    }
  }
  &--icon-r {
    &:not(.#{$r}--icon-only) > .#{$r}__icon {
      margin-inline-start: var(--t-btn-icon-space);
    }
  }
  &__icon {
    &.#{$ct}-enter-from,
    &.#{$ct}-leave-to {
      opacity: 0;
    }
  }
  &.#{$r}--loading {
    cursor: wait;
    opacity: 0.75;
  }
}

@keyframes btn-wave {
  0% {
    opacity: var(--t-btn-wave-opcity);
    box-shadow: 0 0 0.5px 0 var(--t-btn-main-color);
  }
  100% {
    opacity: 0;
    box-shadow: 0 0 0.5px 5px var(--t-btn-main-color);
  }
}
