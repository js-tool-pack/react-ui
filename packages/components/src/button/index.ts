export * from './button.types';
export * from './Button';
export * from './ButtonGroup';
export * from './button.context';
