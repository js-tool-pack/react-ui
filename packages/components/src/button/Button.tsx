import React, { useEffect, useRef, useState } from 'react';
import { ButtonProps } from './button.types';
import {
  CLASS_SIZE_LG,
  CLASS_SIZE_M,
  CLASS_SIZE_SM,
  getComponentClass,
} from '@pkg/shared';
import { getClassNames } from '@tool-pack/basic';
import './Button.module.scss';

const cClass = getComponentClass('button');

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => {
    const {
      className,
      htmlType,
      onClick,
      plain,
      size,
      type,
      children,
      shape,
      ...rest
    } = props;

    const [active, setActive] = useState(false);
    const waveRef = useRef<HTMLElement>(null);

    useEffect(() => {
      const el = waveRef.current;
      if (!active || !el) return;
      const cancel = () => el.removeEventListener('animationend', handler);
      const handler = () => {
        setActive(false);
        cancel();
      };
      el.addEventListener('animationend', handler);

      return cancel;
    }, [active]);

    const clickHandler: React.MouseEventHandler<HTMLButtonElement> = (e) => {
      if (props.disabled) return;
      if (plain !== 'text') setActive(true);
      return onClick?.(e);
    };

    return (
      <button
        {...rest}
        ref={ref}
        type={htmlType}
        onClick={clickHandler}
        className={getClassNames(cClass, `type-${type}`, {
          [CLASS_SIZE_SM]: size === 'small',
          [CLASS_SIZE_M]: size === 'medium',
          [CLASS_SIZE_LG]: size === 'large',
          [`shape-round`]: shape === 'round',
          [`shape-circle`]: shape === 'circle',
          [`plain`]: plain === true,
          [`plain-text`]: plain === 'text',
          [`plain-dashed`]: plain === 'dashed',
          [className as string]: className,
        })}>
        {children}
        <span
          ref={waveRef}
          className={getClassNames('btn-wave', {
            'btn-wave-active': active,
          })}></span>
      </button>
    );
  },
);

Button.displayName = 'Button';
Button.defaultProps = {
  size: 'medium',
  type: 'default',
};
