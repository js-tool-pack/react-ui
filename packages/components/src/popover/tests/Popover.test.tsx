import { Popover } from '..';
import { act, fireEvent, render } from '@testing-library/react';
import { Button } from '~/button';
import { nextTick } from '@tool-pack/basic';

describe('Popover', () => {
  // 模拟 ResizeObserver，ResizeObserver 不存在于 jsdom 中
  const MockObserverInstance: ResizeObserver = {
    observe: jest.fn(),
    unobserve: jest.fn(),
    disconnect: jest.fn(),
  };
  beforeEach(() => {
    global.ResizeObserver = jest
      .fn()
      .mockImplementation(() => MockObserverInstance);
  });

  test('attrs', () => {
    const onClick = jest.fn();
    const { container } = render(
      <Popover
        visible
        attrs={{ className: 'foo', style: { background: '#fff' }, onClick }}
        appendTo={null}>
        <Button>foo bar</Button>
      </Popover>,
    );
    const balloon = container.querySelector('.t-word-balloon');

    expect(balloon).toHaveClass('foo');
    expect(balloon).toHaveStyle({ background: '#fff' });

    expect(onClick).not.toBeCalled();
    fireEvent.click(balloon!);
    expect(onClick).toBeCalled();
  });

  describe('appendTo', () => {
    test('null', () => {
      const { container } = render(
        <Popover visible content={<div>foo bar</div>} appendTo={null}>
          <Button>foo bar</Button>
        </Popover>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    test('undefined(default)', () => {
      const { container } = render(
        <Popover visible content={<div>foo bar</div>} appendTo={undefined}>
          <Button>foo bar</Button>
        </Popover>,
      );
      expect(container.firstChild).toMatchSnapshot();
      expect(document.body).toMatchSnapshot();
    });

    test('fn', () => {
      const { container } = render(
        <div>
          <div className="wrapper"></div>
          <Popover
            visible
            content={<div>foo bar</div>}
            appendTo={() => document.querySelector('.wrapper')!}>
            <Button>foo bar</Button>
          </Popover>
        </div>,
      );
      expect(container.querySelector('.wrapper')).toMatchSnapshot();
    });
  });

  describe('trigger', () => {
    test('focus', () => {
      const { container } = render(
        <Popover trigger="focus" content="focus">
          <input type="text" placeholder="focus 触发" />
        </Popover>,
      );

      expect(document.body).toMatchSnapshot();
      fireEvent.focus(container.querySelector('input')!);
      expect(document.body).toMatchSnapshot();
    });

    test('click', () => {
      const { container } = render(
        <Popover trigger="click" content="click">
          <Button>click触发</Button>
        </Popover>,
      );

      expect(document.body).toMatchSnapshot();
      fireEvent.click(container.querySelector('button')!);
      expect(document.body).toMatchSnapshot();
    });

    test('click disabled', () => {
      const { container } = render(
        <Popover disabled trigger="click" content="click">
          <Button>click触发</Button>
        </Popover>,
      );

      expect(document.body).toMatchSnapshot();
      fireEvent.click(container.querySelector('button')!);
      expect(document.body).toMatchSnapshot();
    });

    test('hover', () => {
      const { container } = render(
        <Popover trigger="hover" content="hover">
          <Button>hover触发</Button>
        </Popover>,
      );

      expect(document.body).toMatchSnapshot();
      fireEvent.mouseEnter(container.querySelector('button')!);
      expect(document.body).toMatchSnapshot();
    });

    test('contextmenu', async () => {
      const { container } = render(
        <Popover trigger="contextmenu" content="contextmenu">
          {/* 不能用 button，因为 contextmenu 事件不能透传给 Button 组件 */}
          {/*<Button className="trigger">鼠标右击触发</Button>*/}
          <div className="trigger">鼠标右击触发</div>
        </Popover>,
      );

      expect(document.body).toMatchSnapshot();
      fireEvent.contextMenu(container.querySelector('.trigger')!);
      await act(() => nextTick());

      expect(document.body).toMatchSnapshot();
    });

    test('arr focus', () => {
      const { container } = render(
        <Popover trigger={['focus', 'click']} content="focus">
          <input type="text" placeholder="focus 触发" />
        </Popover>,
      );
      fireEvent.focus(container.querySelector('input')!);
      expect(document.body).toMatchSnapshot();
    });

    test('arr click', () => {
      const { container } = render(
        <Popover trigger={['focus', 'click']} content="click">
          <Button>click触发</Button>
        </Popover>,
      );
      fireEvent.click(container.querySelector('button')!);
      expect(document.body).toMatchSnapshot();
    });
  });

  test('placement', () => {
    render(
      <Popover visible placement="right-end" content="bar">
        <Button>foo</Button>
      </Popover>,
    );
    expect(document.querySelector('.t-word-balloon')).toHaveClass(
      't-word-balloon--right-end',
    );
  });

  describe('delay', () => {
    test('delay default', () => {
      const { container } = render(
        <Popover trigger="hover" content="hover">
          <Button>hover触发</Button>
        </Popover>,
      );

      expect(document.querySelector('.t-word-balloon')).toBeNull();
      fireEvent.mouseEnter(container.querySelector('button')!);
      expect(document.querySelector('.t-word-balloon')).not.toBeNull();
    });

    test('delay 500', () => {
      jest.useFakeTimers();
      const { container } = render(
        <Popover delay={500} trigger="hover" content="hover">
          <Button>hover触发</Button>
        </Popover>,
      );

      expect(document.querySelector('.t-word-balloon')).toBeNull();
      fireEvent.mouseEnter(container.querySelector('button')!);
      expect(document.querySelector('.t-word-balloon')).toBeNull();
      act(() => jest.advanceTimersByTime(500));
      expect(document.querySelector('.t-word-balloon')).not.toBeNull();
    });

    test('leaveDelay default', () => {
      jest.useFakeTimers();
      const { container } = render(
        <Popover trigger="hover" content="hover">
          <Button>hover触发</Button>
        </Popover>,
      );

      expect(document.querySelector('.t-word-balloon')).toBeNull();
      fireEvent.mouseEnter(container.querySelector('button')!);
      expect(document.querySelector('.t-word-balloon')).not.toBeNull();

      expect(document.querySelector('.t-word-balloon')).not.toHaveClass(
        't-popover-leave-active',
      );
      fireEvent.mouseLeave(document.querySelector('.t-word-balloon')!);
      act(() => jest.advanceTimersByTime(200));
      // todo: 这里暂时测不了，需要给 Transition 组件添加一个默认定时器触发 transitionend 事件
      // expect(document.querySelector('.t-word-balloon')).toHaveClass(
      //   't-popover-leave-active',
      // );
      expect(document.querySelector('.t-word-balloon')).not.toHaveClass(
        't-popover-leave-active',
      );
    });
    test('leaveDelay 500', () => {
      jest.useFakeTimers();
      const { container } = render(
        <Popover trigger="hover" content="hover">
          <Button>hover触发</Button>
        </Popover>,
      );

      expect(document.querySelector('.t-word-balloon')).toBeNull();
      fireEvent.mouseEnter(container.querySelector('button')!);
      expect(document.querySelector('.t-word-balloon')).not.toBeNull();

      expect(document.querySelector('.t-word-balloon')).not.toHaveClass(
        't-popover-leave-active',
      );
      fireEvent.mouseLeave(document.querySelector('.t-word-balloon')!);
      act(() => jest.advanceTimersByTime(500));
      // todo: 这里暂时测不了，需要给 Transition 组件添加一个默认定时器触发 transitionend 事件
      // expect(document.querySelector('.t-word-balloon')).toHaveClass(
      //   't-popover-leave-active',
      // );
      expect(document.querySelector('.t-word-balloon')).not.toHaveClass(
        't-popover-leave-active',
      );
    });
  });
});
