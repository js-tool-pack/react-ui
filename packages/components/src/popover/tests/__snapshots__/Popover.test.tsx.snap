// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover appendTo fn 1`] = `
<div
  class="wrapper"
>
  <div
    class="t-word-balloon t-popover t-word-balloon--top"
    style="top: -10px; left: 0px;"
  >
    <div
      class="t-word-balloon__content"
    >
      <div>
        foo bar
      </div>
    </div>
    <div
      class="t-word-balloon__arrow"
    >
      <svg
        viewBox="-8 -5.5 16 5.515"
      >
        <path
          d="M8-5.5A4 4 180 005.1716-4.3284L1.4142-.5711A2 2 180 01-1.4142-.5711L-5.1716-4.3284A4 4 180 00-8-5.5Z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Popover appendTo null 1`] = `
<button
  class="t-button t-button--type-default t--size-m"
>
  <span>
    foo bar
    <div
      class="t-word-balloon t-popover t-word-balloon--top t-popover-enter-active t-popover-enter-to"
      style="top: -10px; left: 0px;"
    >
      <div
        class="t-word-balloon__content"
      >
        <div>
          foo bar
        </div>
      </div>
      <div
        class="t-word-balloon__arrow"
      >
        <svg
          viewBox="-8 -5.5 16 5.515"
        >
          <path
            d="M8-5.5A4 4 180 005.1716-4.3284L1.4142-.5711A2 2 180 01-1.4142-.5711L-5.1716-4.3284A4 4 180 00-8-5.5Z"
          />
        </svg>
      </div>
    </div>
  </span>
</button>
`;

exports[`Popover appendTo undefined(default) 1`] = `
<button
  class="t-button t-button--type-default t--size-m"
>
  <span>
    foo bar
  </span>
</button>
`;

exports[`Popover appendTo undefined(default) 2`] = `
<body>
  <div>
    <button
      class="t-button t-button--type-default t--size-m"
    >
      <span>
        foo bar
      </span>
    </button>
  </div>
  <div
    class="t-word-balloon t-popover t-word-balloon--top t-popover-enter-active t-popover-enter-to"
    style="top: -10px; left: 0px;"
  >
    <div
      class="t-word-balloon__content"
    >
      <div>
        foo bar
      </div>
    </div>
    <div
      class="t-word-balloon__arrow"
    >
      <svg
        viewBox="-8 -5.5 16 5.515"
      >
        <path
          d="M8-5.5A4 4 180 005.1716-4.3284L1.4142-.5711A2 2 180 01-1.4142-.5711L-5.1716-4.3284A4 4 180 00-8-5.5Z"
        />
      </svg>
    </div>
  </div>
</body>
`;

exports[`Popover trigger arr click 1`] = `
<body>
  <div>
    <button
      class="t-button t-button--type-default t--size-m"
    >
      <span>
        click触发
      </span>
      <span
        class="t-button__wave t-button__wave--active"
      />
    </button>
  </div>
  <div
    class="t-word-balloon t-popover t-word-balloon--top t-popover-enter-active t-popover-enter-to"
    style="top: -10px; left: 0px;"
  >
    <div
      class="t-word-balloon__content"
    >
      click
    </div>
    <div
      class="t-word-balloon__arrow"
    >
      <svg
        viewBox="-8 -5.5 16 5.515"
      >
        <path
          d="M8-5.5A4 4 180 005.1716-4.3284L1.4142-.5711A2 2 180 01-1.4142-.5711L-5.1716-4.3284A4 4 180 00-8-5.5Z"
        />
      </svg>
    </div>
  </div>
</body>
`;

exports[`Popover trigger arr focus 1`] = `
<body>
  <div>
    <input
      placeholder="focus 触发"
      type="text"
    />
  </div>
  <div
    class="t-word-balloon t-popover t-word-balloon--top t-popover-enter-active t-popover-enter-to"
    style="top: -10px; left: 0px;"
  >
    <div
      class="t-word-balloon__content"
    >
      focus
    </div>
    <div
      class="t-word-balloon__arrow"
    >
      <svg
        viewBox="-8 -5.5 16 5.515"
      >
        <path
          d="M8-5.5A4 4 180 005.1716-4.3284L1.4142-.5711A2 2 180 01-1.4142-.5711L-5.1716-4.3284A4 4 180 00-8-5.5Z"
        />
      </svg>
    </div>
  </div>
</body>
`;

exports[`Popover trigger click 1`] = `
<body>
  <div>
    <button
      class="t-button t-button--type-default t--size-m"
    >
      <span>
        click触发
      </span>
    </button>
  </div>
</body>
`;

exports[`Popover trigger click 2`] = `
<body>
  <div>
    <button
      class="t-button t-button--type-default t--size-m"
    >
      <span>
        click触发
      </span>
      <span
        class="t-button__wave t-button__wave--active"
      />
    </button>
  </div>
  <div
    class="t-word-balloon t-popover t-word-balloon--top t-popover-enter-active t-popover-enter-to"
    style="top: -10px; left: 0px;"
  >
    <div
      class="t-word-balloon__content"
    >
      click
    </div>
    <div
      class="t-word-balloon__arrow"
    >
      <svg
        viewBox="-8 -5.5 16 5.515"
      >
        <path
          d="M8-5.5A4 4 180 005.1716-4.3284L1.4142-.5711A2 2 180 01-1.4142-.5711L-5.1716-4.3284A4 4 180 00-8-5.5Z"
        />
      </svg>
    </div>
  </div>
</body>
`;

exports[`Popover trigger click disabled 1`] = `
<body>
  <div>
    <button
      class="t-button t-button--type-default t--size-m"
    >
      <span>
        click触发
      </span>
    </button>
  </div>
</body>
`;

exports[`Popover trigger click disabled 2`] = `
<body>
  <div>
    <button
      class="t-button t-button--type-default t--size-m"
    >
      <span>
        click触发
      </span>
      <span
        class="t-button__wave t-button__wave--active"
      />
    </button>
  </div>
</body>
`;

exports[`Popover trigger contextmenu 1`] = `
<body>
  <div>
    <div
      class="trigger"
    >
      鼠标右击触发
    </div>
  </div>
  <div
    class="t-popover__contextmenu_trigger"
  />
</body>
`;

exports[`Popover trigger contextmenu 2`] = `
<body>
  <div>
    <div
      class="trigger"
    >
      鼠标右击触发
    </div>
  </div>
  <div
    class="t-popover__contextmenu_trigger"
    style="left: 0px; top: 0px;"
  />
  <div
    class="t-word-balloon t-popover t-word-balloon--top t-popover-enter-active t-popover-enter-to"
    style="top: -10px; left: 0px;"
  >
    <div
      class="t-word-balloon__content"
    >
      contextmenu
    </div>
    <div
      class="t-word-balloon__arrow"
    >
      <svg
        viewBox="-8 -5.5 16 5.515"
      >
        <path
          d="M8-5.5A4 4 180 005.1716-4.3284L1.4142-.5711A2 2 180 01-1.4142-.5711L-5.1716-4.3284A4 4 180 00-8-5.5Z"
        />
      </svg>
    </div>
  </div>
</body>
`;

exports[`Popover trigger focus 1`] = `
<body>
  <div>
    <input
      placeholder="focus 触发"
      type="text"
    />
  </div>
</body>
`;

exports[`Popover trigger focus 2`] = `
<body>
  <div>
    <input
      placeholder="focus 触发"
      type="text"
    />
  </div>
  <div
    class="t-word-balloon t-popover t-word-balloon--top t-popover-enter-active t-popover-enter-to"
    style="top: -10px; left: 0px;"
  >
    <div
      class="t-word-balloon__content"
    >
      focus
    </div>
    <div
      class="t-word-balloon__arrow"
    >
      <svg
        viewBox="-8 -5.5 16 5.515"
      >
        <path
          d="M8-5.5A4 4 180 005.1716-4.3284L1.4142-.5711A2 2 180 01-1.4142-.5711L-5.1716-4.3284A4 4 180 00-8-5.5Z"
        />
      </svg>
    </div>
  </div>
</body>
`;

exports[`Popover trigger hover 1`] = `
<body>
  <div>
    <button
      class="t-button t-button--type-default t--size-m"
    >
      <span>
        hover触发
      </span>
    </button>
  </div>
</body>
`;

exports[`Popover trigger hover 2`] = `
<body>
  <div>
    <button
      class="t-button t-button--type-default t--size-m"
    >
      <span>
        hover触发
      </span>
    </button>
  </div>
  <div
    class="t-word-balloon t-popover t-word-balloon--top t-popover-enter-active t-popover-enter-to"
    style="top: -10px; left: 0px;"
  >
    <div
      class="t-word-balloon__content"
    >
      hover
    </div>
    <div
      class="t-word-balloon__arrow"
    >
      <svg
        viewBox="-8 -5.5 16 5.515"
      >
        <path
          d="M8-5.5A4 4 180 005.1716-4.3284L1.4142-.5711A2 2 180 01-1.4142-.5711L-5.1716-4.3284A4 4 180 00-8-5.5Z"
        />
      </svg>
    </div>
  </div>
</body>
`;
